mercury_engine_data_structures/formats/bsmat.py

```diff
@@ -0,0 +1,246 @@
+
+# seems to be a combo of Blend and BlendOp
+# ie "Blend source dest && BlendOp operation"
+blend_state = Struct(
+    enabled = Byte,
+    operation = blend_op,
+    source = blend_mode,
+    dest = blend_mode
+)
+
+# Stencil test
+# https://docs.unity3d.com/Manual/SL-Stencil.html
+stencil_test = Struct(
+    enabled = Byte,
+    mask = Int32ul,
+    ref = Int32ul,
+    fail = stencil_op,
+    success  = stencil_op, # "pass" is from gist, but keywords exist :P
+    depth_fail = stencil_op,
+    depth_success = stencil_op,
+    cmp_mode = compare_mode
+)
+
+# alpha test, seems to be deprecated/not exist in HLSL
+# https://docs.unity3d.com/Manual/SL-AlphaTest.html
+alpha_test = Struct(
+    enabled=Byte,
+    function=compare_mode,
+    ref=Float
+)
+
+# depth test
+depth_state = Struct(
+    depth_test=Byte,
+    depth_write=Byte,
+    depth_mode=compare_mode,
+    z_prepass=Byte
+)
+
+# uniform params, stuff like "vConstant0" or "fAlbedoEmissiveMultiplier"
+uniform_param = Struct(
+    name=StrId,
+    type=Char, # more correctly, a char (f/i/u). but i cant seem to find a FormatField for it. 
+    value=Switch(
+        construct.this.type,
+        {
+            "f": PrefixedArray(Int32ul, Float),
+            "i": PrefixedArray(Int32ul, Int32sl),
+            "u": PrefixedArray(Int32ul, Int32ul)
+        }
+    )
+)
+
+# sampler params, imports texture files typically
+sampler_param = Struct(
+    name=StrId,
+    sampler=StrId,
+    type=StrId,
+    index=Int32ul,
+    filepath=StrId,
+    min_filter=filter_mode,
+    max_filter=filter_mode,
+    mip_filter=filter_mode,
+    cmp_mode=compare_mode,
+    wrap_mode_U=tile_wrap_mode,
+    wrap_mode_V=tile_wrap_mode,
+    border_color=Array(4, Byte),
+    min_lod=Float,
+    lod_bias=Float,
+    anisotropic=Float,
+    max_mip_level=Float,
+    max_anisotropy=Float
+)
+
+shader_stage = Struct(
+    # there was a Int32ul "type" param in the gist here, that seems to be gone now. 
+    uniform_params=PrefixedArray(Int32ul, uniform_param),
+    sampler_params=PrefixedArray(Int32ul, sampler_param)
+)
+
+BSMAT = Struct(
+    _magic = Const(b"MSUR"),
+    _ver = Const(0x00110002, Hex(Int32ul)),
+    name = StrId,
+
+    # Binary shader data
+    type = translucency_type,
+    render_layer = Int32ul,
+    shader_path = StrId,
+    blendstate = blend_state,
+    cullstate = polygon_cull_mode,
+    stencilstate = stencil_test,
+    alphastate = alpha_test,
+    fillmode = fill_mode,
+    depth = depth_state,
+
+    # differs from the gist but its const across all files
+    _const0 = Const(0, Int32ul), # 0
+    _const1 = Const(2, Int32ul), # 2
+    _const2 = Const(0, Int32ul), # 0
+
+    # these also differ from gist and mostly are only in system shaders, from what I can tell
+    extra_uniforms = PrefixedArray(Int32ul, uniform_param),
+    extra_samplers = PrefixedArray(Int32ul, sampler_param),
+
+    # important stuff here!
+    shader_stages = PrefixedArray(Int32ul, shader_stage),
+)
```

Comment:
Is this the end of the file? Add a `_end=construct.Terminated` so we can catch mistakes. Adding a `rest=construct.GreedyBytes` is a good way to cheat if it isn't